### DISCUSSION TITLE
A question on demo_newlib program

### DISCUSSION THREAD

**Discussion by mmcheraghi** [stnolting/neorv32#916] [Q&A] [answered]
Hi @stnolting 
Sorry I couldn't understand the final comment of demo_newlib/main.c. 

// NOTE: exit is highly over-sized as it also includes clean-up functions (destructors), which
  // are not required for bare-metal or RTOS applications... better use the simple 'return' or even better
  // make sure main never returns. Anyway, let's check if 'exit' works.

Would you mind to explain it to me?

**Comment by stnolting** [comment]
Oh, that's a longer story... 😅

`return` just return from the current function. When this is `main()` execution returns to the function that has called `main()`. For embedded systems this is for example the start-up code. In terms of the NEORV32, execution is returned to an endless sleep mode in crt0:

https://github.com/stnolting/neorv32/blob/408269e85f0653a4671caf6b4b5448de39bfda73/sw/common/crt0.S#L171

`exit()`, in contrast, terminates the program wherever it is called. I think this function originates from desktop-like systems as it also performs a lot of other operations (clean-ups) than just terminating the program:
* flush IO streams
* close open files
* free allocated memory
* ...

I think this site gives a pretty nice comparison: https://www.javatpoint.com/difference-between-exit-and-return-in-c

**Reply by mmcheraghi** [reply]
Thanks. The above link was comprehensive. Lastly, What do you mean by "_better use the simple 'return' or even better make sure main never returns_"?

**Reply by stnolting** [reply, marked as answer]
Well, if you want to terminate your `main()` function then just use `return`.

However, when running **bare-metal* code.. what should happen if your code returns / exits? There is no operation system that could handle that.

**Reply by mmcheraghi** [reply]
Thanks
